contracts/modules/VersionManager.sol

```diff
@@ -0,0 +1,250 @@
+        (success, _res) = _wallet.call(abi.encodeWithSignature("invoke(address,uint256,bytes)", _to, _value, _data));
+        if (success && _res.length > 0) { //_res is empty if _wallet is an "old" BaseWallet that can't return output values
+            (_res) = abi.decode(_res, (bytes));
+        } else if (_res.length > 0) {
+            // solhint-disable-next-line no-inline-assembly
+            assembly {
+                returndatacopy(0, 0, returndatasize())
+                revert(0, returndatasize())
+            }
+        } else if (!success) {
+            revert("VM: wallet invoke reverted");
+        }
+    }
+
+    /**
+     * @notice This method delegates the static call to a target feature
+     */
+    fallback() external {
+        uint256 version = walletVersions[msg.sender];
+        address feature = staticCallExecutors[version][msg.sig];
+        require(feature != address(0), "VM: Static call not supported for wallet version");
+
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            calldatacopy(0, 0, calldatasize())
+            let result := staticcall(gas(), feature, 0, calldatasize(), 0, 0)
+            returndatacopy(0, 0, returndatasize())
+            switch result
+            case 0 {revert(0, returndatasize())}
+            default {return (0, returndatasize())}
+        }
+    }
+
+    /* ******* Backward Compatibility with old Storages and BaseWallet *************** */
+
+    /**
+     * @inheritdoc IVersionManager
+     */
+    function setWhitelist(address _wallet, address _target, uint256 _whitelistAfter) external override onlyFeature(_wallet) {
+        transferStorage.setWhitelist(_wallet, _target, _whitelistAfter);
+    }
+
+    /**
+     * @inheritdoc IVersionManager
+     */
+    function setOwner(address _wallet, address _newOwner) external override onlyFeature(_wallet) {
+        IWallet(_wallet).setOwner(_newOwner);
+    }
+
+    /**
+     * @inheritdoc IVersionManager
+     */
+    function addGuardian(address _wallet, address _guardian) external override onlyFeature(_wallet) {
+        guardianStorage.addGuardian(_wallet, _guardian);
+    }
+
+    /**
+     * @inheritdoc IVersionManager
+     */
+    function revokeGuardian(address _wallet, address _guardian) external override onlyFeature(_wallet) {
+        guardianStorage.revokeGuardian(_wallet, _guardian);
+    }
+
+
+    /* ***************** Internal methods ************************* */
+
+
+    function doUpgradeWallet(
+        address _wallet, 
+        address[] memory _featuresToInit
+    ) 
+        internal 
+    {
+        uint256 fromVersion = walletVersions[_wallet];
+        uint256 toVersion = lastVersion;
+        require(fromVersion < toVersion, "VM: Already on last version");
+        walletVersions[_wallet] = toVersion;
+
+        // Setup static call redirection
+        bytes4[] storage sigs = staticCallSignatures[toVersion];
+        for(uint256 i = 0; i < sigs.length; i++) {
```

Comment:
Maybe we could save some gas by introducing a catchall static call for wallets with 1 module (i.e. new wallets) and provided that we slightly update the fallback method of `BaseWallet` : 
```
contract BaseWallet {

    fallback() external payable {
        if(msg.value > 0) {
            emit Received(msg.value, msg.sender, msg.data);
        } else {
            address module = enabled[msg.sig];
            if (module == address(0)) {
                module = enabled[CATCHALL]; // e.g. CATCHALL = 0xffffffff
            }
            require(authorised[module], "BW: must be an authorised module for static call");

            // solhint-disable-next-line no-inline-assembly
            assembly {
                calldatacopy(0, 0, calldatasize())
                let result := staticcall(gas(), module, 0, calldatasize(), 0, 0)
                returndatacopy(0, 0, returndatasize())
                switch result
                case 0 {revert(0, returndatasize())}
                default {return (0, returndatasize())}
            }
        }
    }

}
```

Reply:
Actually this wouldn't work. 

I'm keeping the comment so we can think more about it.